client/trino-client/src/main/java/io/trino/client/spooling/encoding/CompressedQueryDataDecoder.java

```diff
@@ -42,9 +44,47 @@ public QueryDataAccess decode(InputStream stream, DataAttributes metadata)
     {
         Optional<Integer> expectedDecompressedSize = metadata.getOptional(DataAttribute.UNCOMPRESSED_SIZE, Integer.class);
         if (expectedDecompressedSize.isPresent()) {
-            return delegate.decode(decompress(stream, expectedDecompressedSize.get()), metadata);
+            try (ReadOnceByteSource source = new ReadOnceByteSource(stream, expectedDecompressedSize.get())) {
+                return delegate.decode(decompress(source.read(), expectedDecompressedSize.get()), metadata);
+            }
         }
         // Data not compressed - below threshold
         return delegate.decode(stream, metadata);
     }
+
+    private static class ReadOnceByteSource
```

Comment:
Why this name?

Reply:
because it's closing an input stream so you can read all bytes only once